libbpf-rs/src/user_ringbuf.rs

```diff
@@ -0,0 +1,206 @@
+use libc::E2BIG;
+use libc::ENOSPC;
+use std::marker::PhantomData;
+use std::os::fd::AsFd;
+use std::os::fd::AsRawFd;
+use std::os::raw::c_uint;
+use std::os::raw::c_void;
+use std::ptr::null_mut;
+use std::ptr::NonNull;
+use std::sync::Mutex;
+use std::sync::PoisonError;
+
+use crate::AsRawLibbpf;
+use crate::Error;
+use crate::MapHandle;
+use crate::MapType;
+use crate::Result;
+
+/// A mutable reference to a type `T` sample within a [`UserRingBuffer`].
+#[derive(Debug)]
+pub struct UserRingBufferSample<'slf, T> {
+    // A non-null pointer to data of type `T` within the ring buffer.
+    ptr: NonNull<T>,
+
+    // `PhantomData<T>` signals to the Rust compiler the ownership and variance
+    // of `T`. This ensures that `UserRingBufferSample` complies with Rust’s
+    // safety rules regarding lifetimes and drop semantics, despite not storing
+    // `T` directly.
+    _marker: PhantomData<T>,
+
+    // Reference to the owning ring buffer. This is used to discard the sample
+    // if it is not submitted before being dropped.
+    rb: &'slf UserRingBuffer,
+
+    // Track whether the sample has been submitted
+    submitted: bool,
+}
+
+impl<T> UserRingBufferSample<'_, T> {
+    /// Retrieve a mutable reference to the of type `T` within the ring buffer.
+    ///
+    /// You can use this method to modify the data within the ring buffer prior
+    /// to submitting the sample.
+    ///
+    /// # Examples
+    /// ```
+    /// struct MyStruct {
+    ///    field: i32,
+    /// }
+    ///
+    /// let data_ref: &mut MyStruct = sample.as_mut();
+    /// data_ref.field = 42;
+    /// user_ringbuf.submit(sample);
```

Comment:
Yeah, I left it in case the API evolves and returns an error. I looked at the libbpf code for submit and it appears like it just trusts you are giving it a valid sample pointer. I could see that evolving to do some safety checks.

Reply:
Makes sense, let's keep it as it is and return a Result.